source/acquaman/SXRMB/SXRMBXRFScanController.cpp

```diff
@@ -0,0 +1,185 @@
+		scan_->addRawDataSource(new AMRawDataSource(scan_->rawData(), i), true, false);
+	}
+
+	for (int i = 0; i < elements; i++){
+
+		scan_->rawData()->addMeasurement(AMMeasurementInfo(QString("icr%1").arg(i+1), QString("Input count rate %1").arg(i+1), "%", QList<AMAxisInfo>()));
+		scan_->addRawDataSource(new AMRawDataSource(scan_->rawData(), i+elements), false, true);
+	}
+
+	for (int i = 0; i < elements; i++){
+
+		scan_->rawData()->addMeasurement(AMMeasurementInfo(QString("ocr%1").arg(i+1), QString("Output count rate %1").arg(i+1), "%", QList<AMAxisInfo>()));
+		scan_->addRawDataSource(new AMRawDataSource(scan_->rawData(), i+2*elements), false, true);
+	}
+
+	for (int i = 0; i < elements; i++){
+
+		AMDeadTimeAB *temp = new AMDeadTimeAB(QString("Corrected %1").arg(i+1));
+		temp->setInputDataSourcesImplementation(QList<AMDataSource *>() << (AMDataSource *)scan_->rawDataSources()->at(i) << (AMDataSource *)scan_->rawDataSources()->at(i+elements) << (AMDataSource *)scan_->rawDataSources()->at(i+2*elements));
+		scan_->addAnalyzedDataSource(temp, true, false);
+	}
+
+	if (elements > 1){
+
+		AM1DSummingAB *corr = new AM1DSummingAB("corrSum");
+		QList<AMDataSource *> list;
+		for (int i = 0; i < scan_->analyzedDataSourceCount(); i++)
+			list << (AMDataSource *)scan_->analyzedDataSources()->at(i);
+		corr->setInputDataSourcesImplementation(list);
+		scan_->addAnalyzedDataSource(corr, true, false);
+	}
+}
+
+void SXRMBXRFScanController::onProgressUpdate()
+{
+	emit progress(detector_->elapsedTime(), detector_->acquisitionTime());
+}
+
+void SXRMBXRFScanController::onStatusChanged()
+{
+	if (!detector_->isAcquiring())
+		onDetectorAcquisitionFinished();
+}
+
+bool SXRMBXRFScanController::startImplementation()
+{
+	if (!detector_->isConnected() || !detector_->isInitialized()) {
+		QString errorMessage = QString("The bruker detector %1 is not connected or initialized.").arg(detector_->name());
+		AMErrorMon::alert(this, SXRMB::ErrorSXRMBBeamlineXRFDetectorNotConnected, errorMessage, true);
+		return false;
+	}
+
+	connect(detector_, SIGNAL(acquisitionStateChanged(AMDetector::AcqusitionState)), this, SLOT(onStatusChanged()));
+	connect(detector_, SIGNAL(elapsedTimeChanged(double)), this, SLOT(onProgressUpdate()));
+	detector_->acquire();
+
+	setStarted();
+
+	return true;
+}
+
+void SXRMBXRFScanController::onDetectorAcquisitionFinished()
+{
+	disconnect(detector_, SIGNAL(acquisitionStateChanged(AMDetector::AcqusitionState)), this, SLOT(onStatusChanged()));
+	disconnect(detector_, SIGNAL(elapsedTimeChanged(double)), this, SLOT(onProgressUpdate()));
+
+	QVector<double> spectrum = QVector<double>(detector_->size(0));
+
+	for (int i = 0, elements = detector_->elements(); i < elements; i++){
+
+		detector_->rawSpectrumSources().at(i)->values(AMnDIndex(0), AMnDIndex(detector_->size(0)-1), spectrum.data());
+		scan_->rawData()->setValue(AMnDIndex(), i, spectrum.constData());
+		if (detector_->inputCountSourceAt(i))
```

Comment:
Will it? Am I missing something here? The call to `detector_->inputCountSourceAt(i)` actually just retrieves from a QList. The `at()` function is only safe for numbers less than size. Unless I'm missing something the size of that list right now should be 0, so I'm not sure what this is going to do.